src/pkg/packager/publish.go

```diff
@@ -0,0 +1,348 @@
+	copyOpts.Concurrency = p.cfg.PublishOpts.CopyOptions.Concurrency
+	copyOpts.OnCopySkipped = utils.PrintLayerExists
+	copyOpts.PostCopy = utils.PrintLayerExists
+
+	var root ocispec.Descriptor
+
+	root, err = p.publishArtifact(dst, store, descs, copyOpts)
+	if err != nil {
+		// reset the progress bar between attempts
+		dst.ProgressBar.Stop()
+
+		// log the error, the expected error is a 400 manifest invalid
+		message.Debug("ArtifactManifest push failed with the following error, falling back to an ImageManifest push:", err)
+
+		// if the error returned from the push is not an expected error, then return the error
+		if !isManifestUnsupported(err) {
+			return err
+		}
+		root, err = p.publishImage(dst, store, descs, copyOpts)
+		if err != nil {
+			return err
+		}
+	}
+	dst.ProgressBar.Successf("Published %s [%s]", ref, root.MediaType)
+	fmt.Println()
+	flags := ""
+	if config.CommonOptions.Insecure {
+		flags = "--insecure"
+	}
+	message.Info("To inspect/deploy/pull:")
+	message.Infof("zarf package inspect oci://%s/%s %s", ref.Registry, ref.Repository, flags)
+	message.Infof("zarf package deploy oci://%s %s", ref, flags)
+	message.Infof("zarf package pull oci://%s %s", ref, flags)
+	return nil
+}
+
+func (p *Packager) publishArtifact(dst *utils.OrasRemote, store *file.Store, descs []ocispec.Descriptor, copyOpts oras.CopyOptions) (root ocispec.Descriptor, err error) {
+	var total int64
+	for _, desc := range descs {
+		total += desc.Size
+	}
+	packOpts := p.cfg.PublishOpts.PackOptions
+
+	// first attempt to do a ArtifactManifest push
+	root, err = pack(dst.Context, ocispec.MediaTypeArtifactManifest, descs, store, packOpts)
+	if err != nil {
+		return root, err
+	}
+	total += root.Size
+
+	dst.ProgressBar = message.NewProgressBar(total, fmt.Sprintf("Publishing %s:%s", dst.Reference.Repository, dst.Reference.Reference))
+	defer dst.ProgressBar.Stop()
+
+	// attempt to push the artifact manifest
+	_, err = oras.Copy(dst.Context, store, root.Digest.String(), dst, dst.Reference.Reference, copyOpts)
+	return root, err
+}
+
+func (p *Packager) publishImage(dst *utils.OrasRemote, store *file.Store, descs []ocispec.Descriptor, copyOpts oras.CopyOptions) (root ocispec.Descriptor, err error) {
+	var total int64
+	for _, desc := range descs {
+		total += desc.Size
+	}
+	// assumes referrers API is not supported since OCI artifact
+	// media type is not supported
+	dst.SetReferrersCapability(false)
+
+	copyOpts.FindSuccessors = func(ctx context.Context, fetcher content.Fetcher, node ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if content.Equal(node, root) {
+			// skip non-config
+			content, err := content.FetchAll(ctx, fetcher, root)
+			if err != nil {
+				return nil, err
+			}
+			var manifest ocispec.Manifest
+			if err := json.Unmarshal(content, &manifest); err != nil {
+				return nil, err
+			}
+			return []ocispec.Descriptor{manifest.Config}, nil
+		}
+		// config has no successors
+		return nil, nil
+	}
```

Comment:
Not a huge fan of these larger inline functions... It can make it harder to follow where the returns are going.